src/Compilers/CSharp/Portable/Binder/DecisionDagBuilder.cs

```diff
@@ -1785,6 +1795,45 @@ public DagState(ImmutableArray<StateForCase> cases, ImmutableDictionary<BoundDag
             // After the entire graph of DagState objects is complete, we translate each into its Dag node.
             public BoundDecisionDagNode? Dag;
 
+            private DagState()
+            {
+            }
+
+            public void ClearAndFree()
+            {
+                Cases?.Free();
```

Comment:
This is the major benefit.  Making it so that when we let go of the DagState that we put its 'cases array' back in the ArrayBuilder pool to be used by future callers.